### PR TITLE
feat: k8s agent enable disable [DET-9750]

### PR DESCRIPTION
**Pull request by NicholasBlaskey**
## Description

Supports agent enable disable on Kubernetes.

## Test Plan

Should be covered decently well by e2e test

but manual

On a Kubernetes cluster run a bunch of experiments.

Find what agents are in the cluster on by running ```det agent```

(you should require admin to run the following commands)

Disable that node with ```det agent disable nodeName```. Check the experiment was killed and eventually gets rescheduled and completed on a different node.

Repeat but with ```det agent disable nodeName --drain``` and verify the experiment isn't killed.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/pods.go:936]: thoughts on using `metav1.ObjectMeta` instead? and json marshalling it. does it work?

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/pods.go:949]: confusing message for enabled failuire

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/kubernetes_resource_manager.go:59]: plan on doing this todo and removing setClusterID in favor of that? i figure this is just to not have to handle the annoying/impossible error from GetOrCreate... all over

**Review by stoksc — COMMENTED** [review]
would be nice to check our role permissions upfront for features like this and bail out with a nice user facing message, instead of just letting the k8s API call fail. it's easier to reason about how the experience is degraded when we are missing perms, especially in the long run once if this code gets more changes.

File: master/internal/rm/kubernetesrm/pods.go
Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/pods.go:999]: No no no I hate this returning error when it is always nil more than anyone else lol

copy pasta mistaked

Comment (stoksc) [reply master/internal/rm/kubernetesrm/pods.go:999]: lol phew. glad we agree.

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager.go
Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager.go:59]: I added a ticket for it, maybe I'll do it after this lands

spec.go doesn't return errors so it makes it somewhat hard 

FWIW the error isn't impossible we requery GetOrCreate everytime, also we pass clusterID down like a sauasage in like a ton of places where I don't think we probably should

Also blank clusterID in pod.go on restart, but I think the field goes unusued anyway
https://github.com/determined-ai/determined/blob/7c87ee865082ebeb3bbd16013017294077a29ae9/master/internal/rm/kubernetesrm/pods.go#L523-L533

File: master/internal/rm/kubernetesrm/pods.go
Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/pods.go:949]: made message clearer

File: master/internal/rm/kubernetesrm/pods.go
Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/pods.go:936]: That is a good idea

added it for enable but for disable since we need to use the JSON literal null to remove the label, and labels is a ```map[string]string```

I think its interesting how in almost every Go code base you can find things like this

**Comment by NicholasBlaskey** [comment]
> would be nice to check our role permissions upfront for features like this and bail out with a nice user facing message, instead of just letting the k8s API call fail. it's easier to reason about how the experience is degraded when we are missing perms, especially in the long run once if this code gets more changes.

added a check with a helpful message about permissions

File: helm/charts/determined/templates/master-permissions.yaml
Comment (stoksc) [reply helm/charts/determined/templates/master-permissions.yaml:36]: wanna restructure this so all the nodes perms are together?

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [reply master/internal/rm/kubernetesrm/pods.go:937]: oh weird. setting it to "" just leaves it, doesn't remove it? yeah, this is grpc nulls all over again.

File: helm/charts/determined/templates/master-permissions.yaml
Comment (NicholasBlaskey) [reply helm/charts/determined/templates/master-permissions.yaml:36]: done

File: master/internal/rm/kubernetesrm/kubernetes_resource_manager.go
Comment (stoksc) [reply master/internal/rm/kubernetesrm/kubernetes_resource_manager.go]: lol. yeah, let's add a singleton somewhere else eventually, config sounds ok but maybe makes for some weird circular imports.

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [reply master/internal/rm/kubernetesrm/pods.go:1024]: bump

Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/pods.go]: fixed

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/pods.go:493]: nit, pull this into a function with a name that makes it obvious you don't need to think about it ever again after it is written.

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/pods.go:1020]: if you are going to mutate the contents of the node cache, use a lock. or better yet, make a copy and reset it.

the original intent was that the cache was immutable.

File: master/internal/rm/kubernetesrm/pods.go
Comment (stoksc) [inline master/internal/rm/kubernetesrm/pods.go:1037]: better name, and then we don't really need the comment on 1048
```suggestion
	notifiedAllocations := make(map[model.AllocationID]bool)
```

**Review by stoksc — APPROVED** [review]
lgtm modulo final comments

File: master/internal/rm/kubernetesrm/pods.go
Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/pods.go:1020]: https://github.com/determined-ai/determined/blob/84a6612fe68264a0fb8040f46dfc9500a41510d7/master/internal/rm/kubernetesrm/pods.go#L1180C16-L1180C39

confusingly, but also intentionally this doesn't use the node cache I think 

I originally did have the node cache reset the timestamp then query it, but I removed that since in part I think there is no guarantee that our node informer picks up the update by the time we query it?

Maybe I'll add back in the invalidation after just since someone is likely to call ```det a``` after ```det enable/disable```

Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/pods.go:1020]: The intention of this n.Enabled code is just so we don't return wrong results in the proto def

This is proto stuff though that kinda is annoying, since we don't use this result anywhere but still don't want to feel like I'm returning lies in case some person decides to depend on it sometime

Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/pods.go:1020]: Oh now I remember, GetAgents itself doesn't use the cache

Comment (stoksc) [reply master/internal/rm/kubernetesrm/pods.go:1020]: chatted offline, this is a bug. we'll file a ticket and fix it soon.

Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/pods.go:1020]: https://hpe-aiatscale.atlassian.net/browse/DET-9836

**Pull request merged**
